src/coreclr/jit/objectalloc.cpp

```diff
@@ -815,6 +930,96 @@ void ObjectAllocator::RewriteUses()
+
+                if (call->IsHelperCall(m_compiler, CORINFO_HELP_UNBOX))
+                {
+                    JITDUMP("Found unbox helper call [%06u]\n", m_compiler->dspTreeID(call));
+
+                    // See if first arg is possibly a stack allocated box or ref class
+                    // (arg will have been retyped local or local address)
+                    //
+                    CallArg*       secondArg     = call->gtArgs.GetArgByIndex(1);
+                    GenTree* const secondArgNode = secondArg->GetNode();
+
+                    if ((secondArgNode->OperIsLocal() || secondArgNode->OperIs(GT_LCL_ADDR)) &&
+                        !secondArgNode->TypeIs(TYP_REF))
+                    {
+                        const bool                 isForEffect = (user == nullptr) || call->TypeIs(TYP_VOID);
+                        GenTreeLclVarCommon* const lcl         = secondArgNode->AsLclVarCommon();
+
+                        // Rewrite the call to make the box accesses explicit in jitted code.
+                        // user = COMMA(
+                        //           CALL(UNBOX_HELPER_TYPETEST, obj->MethodTable, type),
+                        //           ADD(obj, TARGET_POINTER_SIZE))
+                        //
+                        JITDUMP("Rewriting to invoke box type test helper%s\n", isForEffect ? " for side effect" : "");
+
+                        call->gtCallMethHnd = m_compiler->eeFindHelper(CORINFO_HELP_UNBOX_TYPETEST);
+                        GenTree* const mt   = m_compiler->gtNewMethodTableLookup(lcl, /* onStack */ true);
+                        call->gtArgs.Remove(secondArg);
+                        call->gtArgs.PushBack(m_compiler, NewCallArg::Primitive(mt));
+
+                        if (isForEffect)
+                        {
+                            // call was just for effect, we're done.
+                        }
+                        else
+                        {
+                            GenTree* const lclCopy = m_compiler->gtClone(lcl, /* complexOk */ true);
+                            GenTree* const payloadAddr =
+                                m_compiler->gtNewOperNode(GT_ADD, TYP_BYREF, lclCopy,
+                                                          m_compiler->gtNewIconNode(TARGET_POINTER_SIZE, TYP_I_IMPL));
+                            GenTree* const comma = m_compiler->gtNewOperNode(GT_COMMA, TYP_BYREF, call, payloadAddr);
+                            *use                 = comma;
+
+                            // flag this comma so we can find it later
+                            comma->gtFlags |= GTF_MAKE_CSE;
+                        }
+                    }
+                }
+            }
+            else if (tree->OperIsIndir())
+            {
+                GenTreeIndir* const indir = tree->AsIndir();
+                GenTree* const      addr  = indir->Addr();
+
+                if (addr->OperIs(GT_COMMA) && ((addr->gtFlags & GTF_MAKE_CSE) != 0))
```

Comment:
Can we look for the specific pattern rather than repurposing a flag like this?

Reply:
I suppose we can check if the comma side effect child is the unbox type test helper.

Reply:
done